src/Nethermind/Nethermind.Core/Caching/ClockCache.cs

```diff
@@ -0,0 +1,248 @@
+// SPDX-FileCopyrightText: 2022 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+using Nethermind.Core.Threading;
+
+namespace Nethermind.Core.Caching
+{
+    public sealed class ClockCache<TKey, TValue>
+        where TKey : struct, IEquatable<TKey>
+    {
+        private const int BitShiftPerInt64 = 6;
+        private readonly int _maxCapacity;
+        private readonly ConcurrentDictionary<TKey, LruCacheItem> _cacheMap;
+        private readonly McsLock _lock = new();
+        private readonly string _name;
+        private readonly TKey[] _keys;
+        private readonly long[] _accessedBitmap;
+        private readonly Queue<int> _freeOffsets = new();
```

Comment:
This 3 collections and all the logic around them should be extracted to separate component

Reply:
Moved to base class